File: solutions/159.longest-substring-with-at-most-two-distinct-characters.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring_two_distinct(s)
  hash = Hash.new { |h, k| h[k] = 0 }
  l = 0
  ans = 0
  (0...s.size).each do |r|
    cur = s[r]
    hash[cur] += 1
    while hash.size > 2
      pre = s[l]
      hash[pre] -= 1
      hash.delete(pre) if hash[pre] == 0
      l += 1
    end
    ans = max(ans, r - l + 1)
  end
  ans
end

def max(*vars)=vars.max

File: solutions/159.longest-substring-with-at-most-two-distinct-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstringTwoDistinct = function (s) {
  const hash = new Map();
  let i = 0;
  let ans = 0;
  for (let j = 0; j < s.length; ++j) {
    const c = s[j];
    freq_c = hash.get(c) ?? 0;
    hash.set(c, freq_c + 1);
    while (hash.size > 2) {
      const pre = s[i];
      hash.set(pre, hash.get(pre) - 1);
      if (hash.get(pre) === 0) hash.delete(pre);
      i += 1;
    }
    ans = Math.max(ans, j - i + 1);
  }
  return ans;
};
